### PR TITLE
👌 IMPROVE: Add `myst_heading_slug_func` option

### LINKED ISSUE #331: myst_heading_anchors: Broken for headings with punctuation
**Describe the bug**

With this in `conf.py`:

```py
myst_heading_anchors = 6
```

With the following Markdown:

```md
[Link](#heading-hyphen)
[Link](#example-1-1)

# Heading - Hyphen

# Example 1.1
```

A page is generated with:

```html
<h1>Heading - Hyphen<a class="headerlink" href="#heading-hyphen" title="Permalink to this headline">¶</a></h1>
...
<h1>Example 1.1<a class="headerlink" href="#example-1-1" title="Permalink to this headline">¶</a></h1>
```

However, although the link anchors match the heading anchors, the build warns:

```
WARNING: 'myst' reference target not found: #heading-hyphen
WARNING: 'myst' reference target not found: #example-1-1
```

And the `a` tags are not created (`<span class="xref myst">Link</span>` instead).

**To Reproduce**

Build this repo: https://github.com/jpmckinney/myst-parser-tests

**Expected behavior**

No warning. Links created.

**Environment**

myst-parser 0.13.5

### PR DESCRIPTION
**Pull request by jpmckinney**
closes #331

I need this to avoid having to rewrite hundreds of headings/links, when switching to MyST, which uses a different slugger than Sphinx does (whether migrating from RST or from recommonmark).

### CODE REVIEW

**Comment by jpmckinney** [comment]
I added a commit, because `_resolve_ref_nested` forces `target` to `.lower()`, which is only true when the slug_func is github-slugger style. To be honest, I don't know why the code is there.

I could be wrong, but it seems to be a case of the code being overly helpful and causing an issue. The user should just set correct, lowercase anchors in their Markdown sources.

**Comment by chrisjsewell** [comment]
> To be honest, I don't know why the code is there.

It's taken directly from sphinx, this is what it does for all references

**Comment by chrisjsewell** [comment]
and in turn it does this because docutils lowercases all node ids/names.
So yeh I think removing the lowercasing is not a "trivial" change, and we would have look very carefully if this will not break existing documentation

**Comment by jpmckinney** [comment]
Ah, thanks - I wish Sphinx and docutils internals were better documented :) I'll see what I can do in the extension instead.

**Comment by chrisjsewell** [comment]
>  I wish Sphinx and docutils internals were better documented

lol, you and me both!

File: myst_parser/main.py [hunk 1]
```diff
@@ -186,8 +201,9 @@ def default_parser(config: MdParserConfig) -> MarkdownIt:
         md.use(deflist_plugin)
     if "substitution" in config.enable_extensions:
         md.use(substitution_plugin, *config.sub_delimiters)
-    if config.heading_anchors is not None:
-        md.use(anchors_plugin, max_level=config.heading_anchors)
+    if use_anchors_plugin:
+        max_level = config.heading_anchors if config.heading_anchors is not None else 2
```
Comment (chrisjsewell) [inline myst_parser/main.py:205]: I don't think this is necessary, and will lead to ambiguity. Only `heading_anchors` should be used to set the level.

**Review by chrisjsewell — CHANGES_REQUESTED** [review]
see comment

File: myst_parser/main.py (hunk 1 above)
Comment (jpmckinney) [reply myst_parser/main.py:205]: 2 is the [default](https://github.com/executablebooks/mdit-py-plugins/blob/master/mdit_py_plugins/anchors/index.py#L12), but I can build a kwargs dict instead and not pass max_level if this config isn’t set

Comment (chrisjsewell) [reply myst_parser/main.py:205]: I just don't feel that adding a function should affect whether it is enabled or disabled

Comment (jpmckinney) [reply myst_parser/main.py:205]: I've made a commit to change the behavior.

Comment (chrisjsewell) [reply myst_parser/main.py]: Cheers

**Review by chrisjsewell — APPROVED** [review]
brilliant cheers!

**Comment by kousu** [comment]
Hi! I'm trying to use this but I'm stuck. Think you can help me out a bit?

Context:

```bash
git clone -b myst-issue-4 https://github.com/neuropoly/neuropoly-docs/
cd neuropoly-docs
python -m venv .venv
. .venv/bin/activate
pip install -e .[sphinx]
```

```
(.venv) $ pip freeze | grep myst
myst-parser==0.15.2
```

I inserted some HTML into my section headings in order to be able to style them a bit:

https://github.com/neuropoly/neuropoly-docs/blame/myst-issue-4/team/README.md#L1

this should be a fine thing to do in markdown, but I've realized that sphinx which was built around rST doesn't handle this as I expected. It strips the special characters but *not* the HTML, so I end up with a node named `span-span-team`:

```html
<div class="tex2jax_ignore mathjax_ignore section" id="span-span-team">
<h1><span>👫</span> Team<a class="headerlink" href="#span-span-team" title="Permalink to this headline">¶</a></h1>
```

I dug around and found out about this option and I thought maybe I could use it. I set up this slugger function:

```diff
(.venv) $ git diff
diff --git a/conf.py b/conf.py
index 88927eb..df7c7af 100644
--- a/conf.py
+++ b/conf.py
@@ -81,6 +81,13 @@ html_theme_options = {
 }
 
 
+def slug(s):
+    print("SLUGGING", s)
+    return s
+
+myst_heading_slug_func = slug
+
+
 # On Github's UI, README.{md,txt,rst} are treated as front pages
 # On every other web site, index.{html,php,py,....} are.
 # To make both renders work, glue these two together by symlinks _build/html/**/index.html -> README.html
```

but it crashed:

```bash
(.venv) $ make clean html
Removing everything under '_build'...
Running Sphinx v4.2.0
myst v0.15.2: MdParserConfig(renderer='sphinx', commonmark_only=False, enable_extensions=['dollarmath'], dmath_allow_labels=True, dmath_allow_space=True, dmath_allow_digits=True, dmath_double_inline=False, update_mathjax=True, mathjax_classes='tex2jax_process|mathjax_process|math|output_area', disable_syntax=[], url_schemes=['http', 'https', 'mailto', 'ftp'], heading_anchors=None, heading_slug_func=<function slug at 0x7f271bbc6f70>, html_meta=[], footnote_transition=True, substitutions=[], sub_delimiters=['{', '}'], words_per_minute=200)
building [mo]: targets for 0 po files that are out of date
building [html]: targets for 44 source files that are out of date
updating environment: [new config] 44 added, 0 changed, 0 removed
reading sources... [100%] team/research-associates/taowa-rosetwig                                                                                                                             
looking for now-outdated files... none found
pickling environment... failed

Exception occurred:
  File "/home/kousu/src/neuropoly/website/neuropoly-docs/.venv/lib/python3.9/site-packages/sphinx/builders/__init__.py", line 324, in build
    pickle.dump(self.env, f, pickle.HIGHEST_PROTOCOL)
_pickle.PicklingError: Can't pickle <function slug at 0x7f271bbc6f70>: attribute lookup slug on __main__ failed
The full traceback has been saved in /tmp/sphinx-err-pc2k7pth.log, if you want to report the issue to the developers.
Please also report this if it was a user error, so that a better error message can be provided next time.
A bug report can be filed in the tracker at <https://github.com/sphinx-doc/sphinx/issues>. Thanks!
make: *** [Makefile:20: html] Error 2
```

I was able to run the example from your tests in this PR

```diff
(.venv) [kousu@requiem neuropoly-docs]$ git diff
diff --git a/conf.py b/conf.py
index 88927eb..42e4a38 100644
--- a/conf.py
+++ b/conf.py
@@ -81,6 +81,11 @@ html_theme_options = {
 }
 
 
+from docutils.nodes import make_id
+
+myst_heading_anchors = 2
+myst_heading_slug_func = make_id
+
 # On Github's UI, README.{md,txt,rst} are treated as front pages
 # On every other web site, index.{html,php,py,....} are.
 # To make both renders work, glue these two together by symlinks _build/html/**/index.html -> README.html
```

but it still produced

```
<div class="tex2jax_ignore mathjax_ignore section" id="span-span-team">
```

And confusingly, I tried using the `myst-anchor` example was mentioned in https://myst-parser.readthedocs.io/en/latest/syntax/optional.html#auto-generated-header-anchors but it *disagrees*, instead producing `-team`:

```
(.venv) $ myst-anchors -l 2 team/README.md | grep team
<h1 id="-team"></h1>
```

Since a function defined inside conf.py didn't work, I tried

```
(.venv) $ cat > .venv/lib/python3.9/site-packages/lol.py <<EOF
> def s(s):
    print(s)
    return "myssSYyssysyst"
    #return s
EOF
```

```diff
(.venv) $ git diff

diff --git a/conf.py b/conf.py
index 88927eb..ea5cbc3 100644
--- a/conf.py
+++ b/conf.py
@@ -81,6 +81,11 @@ html_theme_options = {
 }
 
 
+from lol import s
+
+myst_heading_anchors = 2
+myst_heading_slug_func = s
+
 # On Github's UI, README.{md,txt,rst} are treated as front pages
 # On every other web site, index.{html,php,py,....} are.
 # To make both renders work, glue these two together by symlinks _build/html/**/index.html -> README.html
```

This one *did* in fact run, since you can see it printing all the titles:

<details>

```
(.venv) $ make clean html
Removing everything under '_build'...
Running Sphinx v4.2.0
myst v0.15.2: MdParserConfig(renderer='sphinx', commonmark_only=False, enable_extensions=['dollarmath'], dmath_allow_labels=True, dmath_allow_space=True, dmath_allow_digits=True, dmath_double_inline=False, update_mathjax=True, mathjax_classes='tex2jax_process|mathjax_process|math|output_area', disable_syntax=[], url_schemes=['http', 'https', 'mailto', 'ftp'], heading_anchors=2, heading_slug_func=<function s at 0x7ff2ac64a0d0>, html_meta=[], footnote_transition=True, substitutions=[], sub_delimiters=['{', '}'], words_per_minute=200)
building [mo]: targets for 0 po files that are out of date
building [html]: targets for 44 source files that are out of date
updating environment: [new config] 44 added, 0 changed, 0 removed
Welcome to NeuroPoly  2%] README                                                                                                                                                              
✍ Contact Usces... [  4%] contact-us                                                                                                                                                          
Physical Address
Campus Map
Delivery Address
Billing Address
Email
📅 Events & Workshops 6%] events-and-workshops                                                                                                                                                
Calendar
Past Events
🏛 Facilitiesces... [  9%] facilities                                                                                                                                                          
Computer Lab (L5626)
RF Lab (L5613)
MRI Facilities
📣 Job Opportunities 11%] job-opportunities/README                                                                                                                                            
Internshipsrces... [ 13%] job-opportunities/internships/README                                                                                                                                
Labeling Medical Images for A.I. Applicationsnternships/labeling-medical-images-for-a.i.-applications                                                                                         
Medical Image Analysis8%] job-opportunities/internships/medical-image-analysis                                                                                                                
MSc | PhD | PostDoc Positions-opportunities/msc-or-phd-or-postdoc-positions                                                                                                                   
Projects
What profile are we looking for? 
Why you should join us? 
How to apply? 
Project Manager for Software Developmenties/project-manager-for-software-development                                                                                                          
Software Developer [ 25%] job-opportunities/software-developer                                                                                                                                
📚 Publications... [ 27%] publications                                                                                                                                                        
Journal Articles
Book Chapters
Featured Cover Pages
Newspaper/Public Coverage
Theses
🔍 Researchrces... [ 29%] research/README                                                                                                                                                     
Collaborationss... [ 31%] research/collaborations                                                                                                                                             
Data & Scriptss... [ 34%] research/data-scripts                                                                                                                                               
Databases
Image Processing Software research/image-processing-software/README                                                                                                                           
Automatic Spinal Cord Segmentation/image-processing-software/automatic-spinal-cord-segmentation                                                                                               
References
Registration Methods for Spinal Cordmage-processing-software/registration-methods-for-spinal-cord                                                                                             
Robust Multimodal Registration
Template of the Spinal Cordesearch/image-processing-software/template-of-the-spinal-cord                                                                                                      
Mapping Tissue Microstructureearch/mapping-tissue-microstructure/README                                                                                                                       
Biophysical Modelling47%] research/mapping-tissue-microstructure/biophysical-modelling                                                                                                        
Ex Vivo Imaging of Spinal Cord Injurypping-tissue-microstructure/ex-vivo-imaging-of-spinal-cord-injury                                                                                        
Validation Using Histology
Histological Validation of Quantitative MRItissue-microstructure/histological-validation-of-quantitative-mri                                                                                  
In Vivo Measurement of Myelin G-Ratiopping-tissue-microstructure/in-vivo-measurement-of-myelin-g-ratio                                                                                        
Mapping Myeloarchitecture using T2* and Magnetization Transferre/mapping-myeloarchitecture-using-t2-and-magnetization-transfer                                                                
Data Acquisition
Data Processing
Use Multimodal Information to Study Neuronal Architecture
Study the Effect of Tissue Orientation With Respect to B0
References
White Matter Microstructure in Multiple Sclerosis-microstructure/white-matter-microstructure-in-multiple-sclerosis                                                                            
RF & Shim Coil Design61%] research/rf-and-shim-coil-design                                                                                                                                    
RF Coils for MRI
C-Spine Coil for 7T MRI
Real-time shimming
🖥 Softwareurces... [ 63%] software                                                                                                                                                            
Open-Source Projects
👫 Team sources... [ 65%] team/README                                                                                                                                                         
Faculty
Research Associates
Postdoctoral Researchers
PhD Students
MSc Students
Internships
Consultants
Alumni
Aldo Zaimiurces... [ 70%] team/alumni/aldo-zaimi                                                                                                                                              
About
Publications 
Benjamin De Leener [ 72%] team/alumni/benjamin-de-leener                                                                                                                                      
About
Projects
Contact
Publications
Christian S. Perone[ 75%] team/alumni/christian-s.-perone                                                                                                                                     
Publications 
Profiles 
Harris Namirces... [ 77%] team/alumni/harris-nami                                                                                                                                             
Education 
Current Project 
Tanguy Duvalces... [ 79%] team/alumni/tanguy-duval                                                                                                                                            
About
Publications
Julien Cohen-Adad. [ 81%] team/faculty/julien-cohen-adad                                                                                                                                      
Positions
Other Memberships
Contact
Nikola Stikoves... [ 84%] team/faculty/nikola-stikov                                                                                                                                          
Positions
Contact
Agah Karakuzues... [ 86%] team/phd-students/agah-karakuzu                                                                                                                                     
About
Publications (Journal)
Publications (Conference)
Charley Grosces... [ 88%] team/phd-students/charley-gros                                                                                                                                      
Gabriel Mangeat... [ 90%] team/phd-students/gabriel-mangeat                                                                                                                                   
About
Publications 
Ryan Topferrces... [ 93%] team/phd-students/ryan-topfer                                                                                                                                       
Education
Current project
Research Interests
Publications
Contact
Tommy Boshkovski.. [ 95%] team/phd-students/tommy-boshkovski                                                                                                                                  
Education
Current Projects
Contact
Publications
Nibardo Lopez Rios [ 97%] team/research-associates/nibardo-lopez-rios                                                                                                                         
Research Interests
Contact
Taowa Rosetwigs... [100%] team/research-associates/taowa-rosetwig                                                                                                                             
Contact

looking for now-outdated files... none found
pickling environment... done
checking consistency... done
preparing documents... done
writing output... [100%] team/research-associates/taowa-rosetwig                                                                                                                              
/home/kousu/src/neuropoly/website/neuropoly-docs/job-opportunities/software-developer.md:37: WARNING: 'myst' reference target not found: ../team/faculty/julien-cohen-adad.md#email
generating indices... genindex done
writing additional pages... search done
copying images... [100%] .gitbook/assets/tommy.jpeg                                                                                                                                           
copying static files... done
copying extra files... done
dumping search index in English (code: en)... done
dumping object inventory... done
build succeeded, 1 warning.

The HTML pages are in _build/html.
```

</details>

But it didn't actually have any effect: my titles are still all the original text, and the anchors are still `span-span-<whatever>`:

<details>

2021-11-03-152023_874x575_scrot
2021-11-03-152013_864x532_scrot

</details>

Is `myst_heading_slug_func` what I'm supposed to be using? Am I misusing it?

**Comment by sammorley-short** [comment]
I am also getting the same pickling error. Perhaps it's worth adding a test that uses a user-defined slugger function, rather than `docutils.nodes.make_id`?

**Comment by sammorley-short** [comment]
@jpmckinney That worked, thanks! Out of interest though, why does this need to be the case?

**Comment by jpmckinney** [comment]
Sphinx [calls `compile` and `exec` on the `conf.py` file](https://github.com/sphinx-doc/sphinx/blob/f57177de89ff1a154c830558d218c3e334b2b437/sphinx/config.py#L347-L349). As I understand, to pickle a function, it needs to be able to load the module and find the function. Since `conf.py` is not imported, it looks for it in `__main__`, doesn't find it, and "attribute lookup slug on __main__ failed".

**Comment by jpmckinney** [comment]
https://docs.python.org/3/library/pickle.html#what-can-be-pickled-and-unpickled

> Note that functions (built-in and user-defined) are pickled by fully [qualified name](https://docs.python.org/3/glossary.html#term-qualified-name), not by value. [2](https://docs.python.org/3/library/pickle.html#id8) This means that only the function name is pickled, along with the name of the containing module and classes. Neither the function’s code, nor any of its function attributes are pickled. Thus the defining module must be importable in the unpickling environment, and the module must contain the named object, otherwise an exception will be raised.

So, there's a problem if the slug function is in `conf.py`, because `conf.py` is never imported, only `exec`'ed. So Pickle can't determine its fully qualified name. By putting it in another module, its fully qualified name is known.

**Pull request merged**